File: LogisimPlus/LogisimPlus/logisim/circuit/SubcircuitFactory.cs
// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// To produce customized conversions, purchase a Premium Edition license:
// https://www.tangiblesoftwaresolutions.com/product-details/java-to-csharp-converter.html
// ====================================================================================================

using System;
using System.Collections.Generic;

/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

namespace logisim.circuit
{


	using Component = logisim.comp.Component;
	using logisim.data;
	using AttributeSet = logisim.data.AttributeSet;
	using BitWidth = logisim.data.BitWidth;
	using Bounds = logisim.data.Bounds;
	using Direction = logisim.data.Direction;
	using Location = logisim.data.Location;
	using Value = logisim.data.Value;
	using Instance = logisim.instance.Instance;
	using InstanceFactory = logisim.instance.InstanceFactory;
	using InstancePainter = logisim.instance.InstancePainter;
	using InstanceState = logisim.instance.InstanceState;
	using Port = logisim.instance.Port;
	using StdAttr = logisim.instance.StdAttr;
	using Project = logisim.proj.Project;
	using Pin = logisim.std.wiring.Pin;
	using MenuExtender = logisim.tools.MenuExtender;
	using GraphicsUtil = logisim.util.GraphicsUtil;
	using StringGetter = logisim.util.StringGetter;
	using StringUtil = logisim.util.StringUtil;
    using LogisimPlus.Java;

    public class SubcircuitFactory : InstanceFactory
	{
		private class CircuitFeature : StringGetter, MenuExtender, ActionListener
		{
			private readonly SubcircuitFactory outerInstance;

			internal Instance instance;
			internal Project proj;

			public CircuitFeature(SubcircuitFactory outerInstance, Instance instance)
			{
				this.outerInstance = outerInstance;
				this.instance = instance;
			}

			public virtual string get()
			{
				return outerInstance.source.Name;
			}

			public virtual void configureMenu(JPopupMenu menu, Project proj)
			{
				this.proj = proj;
				string name = instance.Factory.DisplayName;
				string text = Strings.get("subcircuitViewItem", name);
				JMenuItem item = new JMenuItem(text);
				item.addActionListener(this);
				menu.add(item);
			}

			public virtual void actionPerformed(ActionEvent e)
			{
				CircuitState superState = proj.CircuitState;
				if (superState == null)
				{
					return;
				}

				CircuitState subState = outerInstance.getSubstate(superState, instance);
				if (subState == null)
				{
					return;
				}
				proj.CircuitState = subState;
			}
		}

		private Circuit source;

		public SubcircuitFactory(Circuit source) : base("", null)
		{
			this.source = source;
			FacingAttribute = StdAttr.FACING;
			DefaultToolTip = new CircuitFeature(this, null);
			InstancePoker = typeof(SubcircuitPoker);
		}

		public virtual Circuit Subcircuit
		{
			get
			{
				return source;
			}
		}

		public override string Name
		{
			get
			{
				return source.Name;
			}
		}

		public override StringGetter DisplayGetter
		{
			get
			{
				return StringUtil.constantGetter(source.Name);
			}
		}

		public override Bounds getOffsetBounds(AttributeSet attrs)
		{
			Direction facing = attrs.getValue(StdAttr.FACING);
			Direction defaultFacing = source.Appearance.Facing;
			Bounds bds = source.Appearance.OffsetBounds;
			return bds.rotate(defaultFacing, facing, 0, 0);
		}

		public override AttributeSet createAttributeSet()
		{
			return new CircuitAttributes(source);
		}

		//
		// methods for configuring instances
		//
		public override void configureNewInstance(Instance instance)
		{
			CircuitAttributes attrs = (CircuitAttributes) instance.AttributeSet;
			attrs.Subcircuit = instance;

			instance.addAttributeListener();
			computePorts(instance);
			// configureLabel(instance); already done in computePorts
		}

		public override void instanceAttributeChanged(Instance instance, Attribute attr)
		{
			if (attr == StdAttr.FACING)
			{
				computePorts(instance);
			}
			else if (attr == CircuitAttributes.LABEL_LOCATION_ATTR)
			{
				configureLabel(instance);
			}
		}

		public override object getInstanceFeature(Instance instance, object key)
		{
			if (key == typeof(MenuExtender))
			{
				return new CircuitFeature(this, instance);
			}
			return base.getInstanceFeature(instance, key);
		}

		internal virtual void computePorts(Instance instance)
		{
			Direction facing = instance.getAttributeValue(StdAttr.FACING);
			IDictionary<Location, Instance> portLocs = source.Appearance.getPortOffsets(facing);
			Port[] ports = new Port[portLocs.Count];
			Instance[] pins = new Instance[portLocs.Count];
			int i = -1;
			foreach (KeyValuePair<Location, Instance> portLoc in portLocs.SetOfKeyValuePairs())
			{
				i++;
				Location loc = portLoc.Key;
				Instance pin = portLoc.Value;
				string type = Pin.FACTORY.isInputPin(pin) ? Port.INPUT : Port.OUTPUT;
				BitWidth width = pin.getAttributeValue(StdAttr.WIDTH);
				ports[i] = new Port(loc.X, loc.Y, type, width);
				pins[i] = pin;

				string label = pin.getAttributeValue(StdAttr.LABEL);
				if (!string.ReferenceEquals(label, null) && label.Length > 0)
				{
					ports[i].setToolTip(StringUtil.constantGetter(label));
				}
			}

			CircuitAttributes attrs = (CircuitAttributes) instance.AttributeSet;
			attrs.PinInstances = pins;
			instance.setPorts(ports);
			instance.recomputeBounds();
			configureLabel(instance); // since this affects the circuit's bounds
		}

		private void configureLabel(Instance instance)
		{
			Bounds bds = instance.Bounds;
			Direction loc = instance.getAttributeValue(CircuitAttributes.LABEL_LOCATION_ATTR);

			int x = bds.X + bds.Width / 2;
			int y = bds.Y + bds.Height / 2;
			int ha = GraphicsUtil.H_CENTER;
			int va = GraphicsUtil.V_CENTER;
			if (loc == Direction.East)
			{
				x = bds.X + bds.Width + 2;
				ha = GraphicsUtil.H_LEFT;
			}
			else if (loc == Direction.West)
			{
				x = bds.X - 2;
				ha = GraphicsUtil.H_RIGHT;
			}
			else if (loc == Direction.South)
			{
				y = bds.Y + bds.Height + 2;
				va = GraphicsUtil.V_TOP;
			}
			else
			{
				y = bds.Y - 2;
				va = GraphicsUtil.V_BASELINE;
			}
			instance.setTextField(StdAttr.LABEL, StdAttr.LABEL_FONT, x, y, ha, va);
		}

		//
		// propagation-oriented methods
		//
		public virtual CircuitState getSubstate(CircuitState superState, Instance instance)
		{
			return getSubstate(createInstanceState(superState, instance));
		}

		public virtual CircuitState getSubstate(CircuitState superState, Component comp)
		{
			return getSubstate(createInstanceState(superState, comp));
		}

		private CircuitState getSubstate(InstanceState instanceState)
		{
			CircuitState subState = (CircuitState) instanceState.Data;
			if (subState == null)
			{
				subState = new CircuitState(instanceState.Project, source);
				instanceState.Data = subState;
				instanceState.fireInvalidated();
			}
			return subState;
		}

		public override void propagate(InstanceState superState)
		{
			CircuitState subState = getSubstate(superState);

			CircuitAttributes attrs = (CircuitAttributes) superState.AttributeSet;
			Instance[] pins = attrs.PinInstances;
			for (int i = 0; i < pins.Length; i++)
			{
				Instance pin = pins[i];
				InstanceState pinState = subState.getInstanceState(pin);
				if (Pin.FACTORY.isInputPin(pin))
				{
					Value newVal = superState.getPort(i);
					Value oldVal = Pin.FACTORY.getValue(pinState);
					if (!newVal.Equals(oldVal))
					{
						Pin.FACTORY.setValue(pinState, newVal);
						Pin.FACTORY.propagate(pinState);
					}
				}
				else
				{ // it is output-only
					Value val = pinState.getPort(0);
					superState.setPort(i, val, 1);
				}
			}
		}

		//
		// user interface features
		//
		public override void paintGhost(InstancePainter painter)
		{
			Graphics g = painter.Graphics;
			Color fg = g.getColor();
			int v = fg.getRed() + fg.getGreen() + fg.getBlue();
			Composite oldComposite = null;
			if (g is Graphics2D && v > 50)
			{
				oldComposite = ((Graphics2D) g).getComposite();
				Composite c = AlphaComposite.getInstance(AlphaComposite.SRC_OVER, 0.5f);
				((Graphics2D) g).setComposite(c);
			}
			paintBase(painter, g);
			if (oldComposite != null)
			{
				((Graphics2D) g).setComposite(oldComposite);
			}
		}

		public override void paintInstance(InstancePainter painter)
		{
			paintBase(painter, painter.Graphics);
			painter.drawPorts();
		}

		private void paintBase(InstancePainter painter, Graphics g)
		{
			CircuitAttributes attrs = (CircuitAttributes) painter.AttributeSet;
			Direction facing = attrs.Facing;
			Direction defaultFacing = source.Appearance.Facing;
			Location loc = painter.Location;
			g.translate(loc.X, loc.Y);
			source.Appearance.paintSubcircuit(g, facing);
			drawCircuitLabel(painter, getOffsetBounds(attrs), facing, defaultFacing);
			g.translate(-loc.X, -loc.Y);
			painter.drawLabel();
		}

		private void drawCircuitLabel(InstancePainter painter, Bounds bds, Direction facing, Direction defaultFacing)
		{
			AttributeSet staticAttrs = source.StaticAttributes;
			string label = staticAttrs.getValue(CircuitAttributes.CIRCUIT_LABEL_ATTR);
			if (!string.ReferenceEquals(label, null) && !label.Equals(""))
			{
				Direction up = staticAttrs.getValue(CircuitAttributes.CIRCUIT_LABEL_FACING_ATTR);
				Font font = staticAttrs.getValue(CircuitAttributes.CIRCUIT_LABEL_FONT_ATTR);

				int back = label.IndexOf('\\');
				int lines = 1;
				bool backs = false;
				while (back >= 0 && back <= label.Length - 2)
				{
					char c = label[back + 1];
					if (c == 'n')
					{
						lines++;
					}
					else if (c == '\\')
					{
						backs = true;
					}
					back = label.IndexOf('\\', back + 2);
				}

				int x = bds.X + bds.Width / 2;
				int y = bds.Y + bds.Height / 2;
				JGraphics g = painter.Graphics;
				double angle = Math.PI / 2 - (up.toRadians() - defaultFacing.toRadians()) - facing.toRadians();
				if (Math.Abs(angle) > 0.01)
				{
					g.rotate(angle, x, y);
				}
				g.setFont(font);
				if (lines == 1 && !backs)
				{
					GraphicsUtil.drawCenteredText(g, label, x, y);
				}
				else
				{
                    SizeF labelSize = g.measureString(label);
                    int height = labelSize.Height;
					y = y - (height * lines - labelSize.Height) / 2 + labelSize.Height;
					back = label.IndexOf('\\');
					while (back >= 0 && back <= label.Length - 2)
					{
						char c = label[back + 1];
						if (c == 'n')
						{
							string line = label.Substring(0, back);
							GraphicsUtil.drawText(g, line, x, y, GraphicsUtil.H_CENTER, GraphicsUtil.V_BASELINE);
							y += height;
							label = label.Substring(back + 2);
							back = label.IndexOf('\\');
						}
						else if (c == '\\')
						{
							label = label.Substring(0, back) + label.Substring(back + 1);
							back = label.IndexOf('\\', back + 1);
						}
						else
						{
							back = label.IndexOf('\\', back + 2);
						}
					}
					GraphicsUtil.drawText(g, label, x, y, GraphicsUtil.H_CENTER, GraphicsUtil.V_BASELINE);
				}
				g.dispose();
			}
		}

		/*
		 * TODO public String getToolTip(ComponentUserEvent e) { return
		 * StringUtil.format(Strings.get("subcircuitCircuitTip"), source.getDisplayName()); }
		 */
	}

}

File: logisim/app/src/main/java/logisim/circuit/SubcircuitFactory.java
/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

package logisim.circuit;

import java.awt.AlphaComposite;
import java.awt.Color;
import java.awt.Composite;
import java.awt.Font;
import java.awt.FontMetrics;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.event.ActionListener;
import java.awt.event.ActionEvent;
import java.util.Map;
import java.util.Map.Entry;

import javax.swing.JPopupMenu;
import javax.swing.JMenuItem;

import logisim.comp.Component;
import logisim.data.Attribute;
import logisim.data.AttributeSet;
import logisim.data.BitWidth;
import logisim.data.Bounds;
import logisim.data.Direction;
import logisim.data.Location;
import logisim.data.WireValue.WireValue;
import logisim.instance.Instance;
import logisim.instance.InstanceFactory;
import logisim.instance.InstancePainter;
import logisim.instance.InstanceState;
import logisim.instance.Port;
import logisim.instance.StdAttr;
import logisim.proj.Project;
import logisim.std.wiring.Pin;
import logisim.tools.MenuExtender;
import logisim.util.GraphicsUtil;
import logisim.util.StringGetter;
import logisim.util.StringUtil;

public class SubcircuitFactory extends InstanceFactory {
	private class CircuitFeature implements StringGetter, MenuExtender, ActionListener {
		private Instance instance;
		private Project proj;

		public CircuitFeature(Instance instance) {
			this.instance = instance;
		}

		public String get() {
			return source.getName();
		}

		public void configureMenu(JPopupMenu menu, Project proj) {
			this.proj = proj;
			String name = instance.getFactory().getDisplayName();
			String text = Strings.get("subcircuitViewItem", name);
			JMenuItem item = new JMenuItem(text);
			item.addActionListener(this);
			menu.add(item);
		}

		public void actionPerformed(ActionEvent e) {
			CircuitState superState = proj.getCircuitState();
			if (superState == null)
				return;

			CircuitState subState = getSubstate(superState, instance);
			if (subState == null)
				return;
			proj.setCircuitState(subState);
		}
	}

	private Circuit source;

	public SubcircuitFactory(Circuit source) {
		super("", null);
		this.source = source;
		setFacingAttribute(StdAttr.FACING);
		setDefaultToolTip(new CircuitFeature(null));
		setInstancePoker(SubcircuitPoker.class);
	}

	public Circuit getSubcircuit() {
		return source;
	}

	@Override
	public String getName() {
		return source.getName();
	}

	@Override
	public StringGetter getDisplayGetter() {
		return StringUtil.constantGetter(source.getName());
	}

	@Override
	public Bounds getOffsetBounds(AttributeSet attrs) {
		Direction facing = attrs.getValue(StdAttr.FACING);
		Direction defaultFacing = source.getAppearance().getFacing();
		Bounds bds = source.getAppearance().getOffsetBounds();
		return bds.rotate(defaultFacing, facing, 0, 0);
	}

	@Override
	public AttributeSet createAttributeSet() {
		return new CircuitAttributes(source);
	}

	//
	// methods for configuring instances
	//
	@Override
	public void configureNewInstance(Instance instance) {
		CircuitAttributes attrs = (CircuitAttributes) instance.getAttributeSet();
		attrs.setSubcircuit(instance);

		instance.addAttributeListener();
		computePorts(instance);
		// configureLabel(instance); already done in computePorts
	}

	@Override
	public void instanceAttributeChanged(Instance instance, Attribute<?> attr) {
		if (attr == StdAttr.FACING) computePorts(instance);
		else if (attr == CircuitAttributes.LABEL_LOCATION_ATTR) configureLabel(instance);
	}

	@Override
	public Object getInstanceFeature(Instance instance, Object key) {
		if (key == MenuExtender.class)
			return new CircuitFeature(instance);
		return super.getInstanceFeature(instance, key);
	}

	void computePorts(Instance instance) {
		Direction facing = instance.getAttributeValue(StdAttr.FACING);
		Map<Location, Instance> portLocs = source.getAppearance().getPortOffsets(facing);
		Port[] ports = new Port[portLocs.size()];
		Instance[] pins = new Instance[portLocs.size()];
		int i = -1;
		for (Entry<Location, Instance> portLoc : portLocs.entrySet()) {
			i++;
			Location loc = portLoc.getKey();
			Instance pin = portLoc.getValue();
			String type = Pin.FACTORY.isInputPin(pin) ? Port.INPUT : Port.OUTPUT;
			BitWidth width = pin.getAttributeValue(StdAttr.WIDTH);
			ports[i] = new Port(loc.x(), loc.y(), type, width);
			pins[i] = pin;

			String label = pin.getAttributeValue(StdAttr.LABEL);
			if (label != null && !label.isEmpty()) ports[i].setToolTip(StringUtil.constantGetter(label));
		}

		CircuitAttributes attrs = (CircuitAttributes) instance.getAttributeSet();
		attrs.setPinInstances(pins);
		instance.setPorts(ports);
		instance.recomputeBounds();
		configureLabel(instance); // since this affects the circuit's bounds
	}

	private void configureLabel(Instance instance) {
		Bounds bds = instance.getBounds();
		Direction loc = instance.getAttributeValue(CircuitAttributes.LABEL_LOCATION_ATTR);

		int x = bds.getX() + bds.getWidth() / 2;
		int y = bds.getY() + bds.getHeight() / 2;
		int ha = GraphicsUtil.H_CENTER;
		int va = GraphicsUtil.V_CENTER;
		if (loc == Direction.East) {
			x = bds.getX() + bds.getWidth() + 2;
			ha = GraphicsUtil.H_LEFT;
		} else if (loc == Direction.West) {
			x = bds.getX() - 2;
			ha = GraphicsUtil.H_RIGHT;
		} else if (loc == Direction.South) {
			y = bds.getY() + bds.getHeight() + 2;
			va = GraphicsUtil.V_TOP;
		} else {
			y = bds.getY() - 2;
			va = GraphicsUtil.V_BASELINE;
		}
		instance.setTextField(StdAttr.LABEL, StdAttr.LABEL_FONT, x, y, ha, va);
	}

	//
	// propagation-oriented methods
	//
	public CircuitState getSubstate(CircuitState superState, Instance instance) {
		return getSubstate(createInstanceState(superState, instance));
	}

	public CircuitState getSubstate(CircuitState superState, Component comp) {
		return getSubstate(createInstanceState(superState, comp));
	}

	private CircuitState getSubstate(InstanceState instanceState) {
		CircuitState subState = (CircuitState) instanceState.getData();
		if (subState == null) {
			subState = new CircuitState(instanceState.getProject(), source);
			instanceState.setData(subState);
			instanceState.fireInvalidated();
		}
		return subState;
	}

	@Override
	public void propagate(InstanceState superState) {
		CircuitState subState = getSubstate(superState);

		CircuitAttributes attrs = (CircuitAttributes) superState.getAttributeSet();
		Instance[] pins = attrs.getPinInstances();
		for (int i = 0; i < pins.length; i++) {
			Instance pin = pins[i];
			InstanceState pinState = subState.getInstanceState(pin);
			if (Pin.FACTORY.isInputPin(pin)) {
				WireValue newVal = superState.getPort(i);
				WireValue oldVal = Pin.FACTORY.getValue(pinState);
				if (!newVal.equals(oldVal)) {
					Pin.FACTORY.setValue(pinState, newVal);
					Pin.FACTORY.propagate(pinState);
				}
			} else { // it is output-only
				WireValue val = pinState.getPort(0);
				superState.setPort(i, val, 1);
			}
		}
	}

	//
	// user interface features
	//
	@Override
	public void paintGhost(InstancePainter painter) {
		Graphics g = painter.getGraphics();
		Color fg = g.getColor();
		int v = fg.getRed() + fg.getGreen() + fg.getBlue();
		Composite oldComposite = null;
		if (g instanceof Graphics2D && v > 50) {
			oldComposite = ((Graphics2D) g).getComposite();
			Composite c = AlphaComposite.getInstance(AlphaComposite.SRC_OVER, 0.5f);
			((Graphics2D) g).setComposite(c);
		}
		paintBase(painter, g);
		if (oldComposite != null) ((Graphics2D) g).setComposite(oldComposite);
	}

	@Override
	public void paintInstance(InstancePainter painter) {
		paintBase(painter, painter.getGraphics());
		painter.drawPorts();
	}

	private void paintBase(InstancePainter painter, Graphics g) {
		CircuitAttributes attrs = (CircuitAttributes) painter.getAttributeSet();
		Direction facing = attrs.getFacing();
		Direction defaultFacing = source.getAppearance().getFacing();
		Location loc = painter.getLocation();
		g.translate(loc.x(), loc.y());
		source.getAppearance().paintSubcircuit(g, facing);
		drawCircuitLabel(painter, getOffsetBounds(attrs), facing, defaultFacing);
		g.translate(-loc.x(), -loc.y());
		painter.drawLabel();
	}

	private void drawCircuitLabel(InstancePainter painter, Bounds bds, Direction facing, Direction defaultFacing) {
		AttributeSet staticAttrs = source.getStaticAttributes();
		String label = staticAttrs.getValue(CircuitAttributes.CIRCUIT_LABEL_ATTR);
		if (label != null && !label.isEmpty()) {
			Direction up = staticAttrs.getValue(CircuitAttributes.CIRCUIT_LABEL_FACING_ATTR);
			Font font = staticAttrs.getValue(CircuitAttributes.CIRCUIT_LABEL_FONT_ATTR);

			int back = label.indexOf('\\');
			int lines = 1;
			boolean backs = false;
			while (back >= 0 && back <= label.length() - 2) {
				char c = label.charAt(back + 1);
				if (c == 'n')
					lines++;
				else if (c == '\\')
					backs = true;
				back = label.indexOf('\\', back + 2);
			}

			int x = bds.getX() + bds.getWidth() / 2;
			int y = bds.getY() + bds.getHeight() / 2;
			Graphics g = painter.getGraphics().create();
			double angle = Math.PI / 2 - (up.toRadians() - defaultFacing.toRadians()) - facing.toRadians();
			if (g instanceof Graphics2D g2 && Math.abs(angle) > 0.01) g2.rotate(angle, x, y);
			g.setFont(font);
			if (lines == 1 && !backs) GraphicsUtil.drawCenteredText(g, label, x, y);
			else {
				FontMetrics fm = g.getFontMetrics();
				int height = fm.getHeight();
				y = y - (height * lines - fm.getLeading()) / 2 + fm.getAscent();
				back = label.indexOf('\\');
				while (back >= 0 && back <= label.length() - 2) {
					char c = label.charAt(back + 1);
					if (c == 'n') {
						String line = label.substring(0, back);
						GraphicsUtil.drawText(g, line, x, y, GraphicsUtil.H_CENTER, GraphicsUtil.V_BASELINE);
						y += height;
						label = label.substring(back + 2);
						back = label.indexOf('\\');
					} else if (c == '\\') {
						label = label.substring(0, back) + label.substring(back + 1);
						back = label.indexOf('\\', back + 1);
					} else back = label.indexOf('\\', back + 2);
				}
				GraphicsUtil.drawText(g, label, x, y, GraphicsUtil.H_CENTER, GraphicsUtil.V_BASELINE);
			}
			g.dispose();
		}
	}

	/*
	 * TODO public String getToolTip(ComponentUserEvent e) { return
	 * StringUtil.format(Strings.get("subcircuitCircuitTip"), source.getDisplayName()); }
	 */
}
